six/modules/c++/six.sicd/source/SCPCOA.cpp

```diff
@@ -59,3 +68,236 @@ bool SCPCOA::operator==(const SCPCOA& rhs) const
         azimAngle == rhs.azimAngle &&
         layoverAngle == rhs.layoverAngle);
 }
+
+double SCPCOA::derivedSCPTime(const Grid& grid) const
+{
+    return grid.timeCOAPoly[0][0];
+}
+
+double SCPCOA::derivedSlantRange(const Vector3& scp) const
+{
+    return (scp - arpPos).norm();
+}
+
+double SCPCOA::derivedGroundRange(const Vector3& scp) const
+{
+    return scp.norm() * arpPos.angle(scp);
+}
+
+Vector3 SCPCOA::slantPlaneNormal(const Vector3& scp) const
+{
+    return scene::SceneGeometry(arpVel, arpPos, scp).getSlantPlaneZ();
+}
+
+double SCPCOA::derivedTwistAngle(const scene::SceneGeometry& geometry) const
+{
+    // Angle from +GPY axis to the +SPY axis in plane of incidence
+    Vector3 uGPY = cross(geometry.getGroundPlaneNormal(),
+            -geometry.getGroundRange().unit());
+    return -1 * std::asin(uGPY.dot(geometry.getSlantPlaneZ())) *
+        math::Constants::RADIANS_TO_DEGREES;
+}
+
+void SCPCOA::fillDerivedFields(const GeoData& geoData,
+        const Grid& grid,
+        const Position& position)
+{
+    if (Init::isUndefined<double>(scpTime) &&
+        !Init::isUndefined<Poly2D>(grid.timeCOAPoly))
+    {
+        scpTime = derivedSCPTime(grid);
+    }
+    if (!Init::isUndefined<PolyXYZ>(position.arpPoly) &&
+        !Init::isUndefined<double>(scpTime))
+    {
+        std::vector<Vector3> vectors = derivedArpVectors(position);
+        if (Init::isUndefined<Vector3>(arpPos))
+        {
+            arpPos = vectors[0];
+        }
+        if (Init::isUndefined<Vector3>(arpVel))
+        {
+            arpVel = vectors[1];
+        }
+        if (Init::isUndefined<Vector3>(arpAcc))
+        {
+            arpAcc = vectors[2];
+        }
+    }
+
+    scene::SceneGeometry geometry(arpVel, arpPos, geoData.scp.ecf);
+
+    const Vector3& scp = geoData.scp.ecf;
+    if (sideOfTrack == SideOfTrackType::NOT_SET)
+    {
+        sideOfTrack = geometry.getSideOfTrack();
+    }
+    if (Init::isUndefined<double>(slantRange))
+    {
+        slantRange = derivedSlantRange(scp);
+    }
+    if (Init::isUndefined<double>(groundRange))
+    {
+        groundRange = derivedGroundRange(scp);
+    }
+    if (Init::isUndefined<double>(dopplerConeAngle))
+    {
+        dopplerConeAngle = geometry.getDopplerConeAngle();
+    }
+    if (Init::isUndefined<double>(grazeAngle))
+    {
+        grazeAngle = std::fabs(geometry.getETPGrazingAngle());
```

Comment:
Just use `std::abs()` rather than `std::fabs()` here and below - it's overloaded so that it'll do what you want for a `double`